pkg/reporting/client.go

```diff
@@ -0,0 +1,143 @@
+package reporting
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.uber.org/zap"
+)
+
+type Client[E any] struct {
+	Name            string
+	Base            BaseClient
+	BaseConfig      BaseClientConfig
+	GenerateTraceID func() string
+	SerializeBatch  func(events []E) ([]byte, SimpleError)
+}
+
+// BaseClient is the shared lower-level interface to send the processed data somewhere.
+//
+// See S3Client, AzureBlobClient, and HTTPClient.
+type BaseClient interface {
+	LogFields() zap.Field
+	Send(ctx context.Context, payload []byte, traceID string) SimpleError
+}
+
+type BaseClientConfig struct {
+	PushEverySeconds          uint `json:"pushEverySeconds"`
+	PushRequestTimeoutSeconds uint `json:"pushRequestTimeoutSeconds"`
+	MaxBatchSize              uint `json:"maxBatchSize"`
+}
+
+// SimpleError is an extension of the standard 'error' interface that provides a safe-for-metrics
+// string representing the error.
+type SimpleError interface {
```

Comment:
I chose not to have a struct here as the general advice with Go is to avoid returning concrete error types; ref https://go.dev/doc/faq#nil_error

(and, I also wanted it usage of it to be as close to `error` as possible, just with an added method)

Reply:
Changed the name in f60856939cb1faf76f9a3de0901a805b58f1d37d